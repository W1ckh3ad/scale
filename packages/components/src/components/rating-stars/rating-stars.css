:host {
  --custom-outline-spread: 0.25rem;
}

sl-icon {
  height: 100%;
}

.rating {
  position: relative;
  display: inline-flex;
  border-radius: 0.25rem;
  /* border: 5px solid black; */
  margin-left: var(--custom-outline-spread);
}

.rating:focus {
  outline: none;
  /* box-shadow: 0 0 0 var( --custom-outline-spread) rgb(0, 238, 255); */
}

.rating-focus:focus {
  box-shadow: 0 0 0 var(--custom-outline-spread) rgb(0, 238, 255);
}

.rating__symbols {
  display: inline-flex;
  position: relative;
  cursor: pointer;
  padding: 0.1rem 0.125rem 0px 0.125rem;
}

.rating__symbols--indicator {
  position: absolute;
  top: 0;
  left: 0;
}

.rating__symbol {
  transition: 100ms transform;
  margin-right: var(--star-spacing, 4px);
}

.rating__symbol:last-child {
  padding-right: 0;
}

.rating__symbol--hover {
  transform: scale(1.2);
}

.rating__symbol--hover .star-filled {
  fill: var(--color-secondary, #e20074);
}

.rating--interactive .rating__symbols,
.rating--disabled .rating__symbols {
  cursor: default;
}

.rating--interactive .rating__symbol--hover,
.rating--disabled .rating__symbol--hover {
  transform: none;
}

.rating--disabled .rating__symbols {
  opacity: 50%;
}
