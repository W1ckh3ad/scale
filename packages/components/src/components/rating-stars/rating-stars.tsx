import { Component, h, Prop } from '@stencil/core';
import { clamp, handleListeners } from './utils/utils';

@Component({
  tag: 'scale-rating-stars',
  styleUrl: 'rating-stars.css',
  shadow: true,
})
export class RatingStars {
  element: HTMLElement;
  @Prop({ mutable: true }) hoverValue = 0;
  @Prop({ mutable: true }) isHovering = false;
  @Prop({ mutable: true }) numOfStars = 5;
  @Prop({ mutable: true }) rating = 0;
  @Prop({ mutable: true }) small = false;
  @Prop({ mutable: true }) interactive = true;
  @Prop({ mutable: true }) disabled = false;
  @Prop() ariaLabel: string;
  @Prop() precision = 1;
  @Prop() getSymbolBlank = () => '<star-icon></star-icon>';
  @Prop() getSymbolFilled = () => `<star-icon colored></star-icon>`;

  connectedCallback() {
    this.handleMouseEnter = this.handleMouseEnter.bind(this);
    this.handleMouseLeave = this.handleMouseLeave.bind(this);
    this.handleMouseMove = this.handleMouseMove.bind(this);
    this.handleMouseClick = this.handleMouseClick.bind(this);
    this.handleKeyDown = this.handleKeyDown.bind(this);
  }

  componentDidLoad() {
    handleListeners(this.element, 'addListeners');
  }

  disconnectedCallback() {
    handleListeners(this.element, 'removeListeners');
  }

  handleMouseLeave() {
    this.isHovering = false;
    return false;
  }

  handleMouseEnter() {
    if (this.disabled || !this.interactive) {
      return;
    }
    this.isHovering = true;
    return true;
  }

  handleMouseMove(event: MouseEvent) {
    this.hoverValue = this.getValueFromMousePosition(event);
    // console.log(this.hoverValue);
  }

  handleMouseClick(event: MouseEvent) {
    if (this.disabled || !this.interactive) {
      return;
    }

    this.isHovering = false;
    this.rating =
      this.rating === this.hoverValue
        ? 0
        : this.getValueFromMousePosition(event);
  }

  handleKeyDown(event: KeyboardEvent) {
    if (this.disabled || !this.interactive) {
      return;
    }

    if (event.key === 'ArrowRight') {
      const ratingPlus = this.rating + this.precision;
      this.rating = clamp(ratingPlus, 0, this.numOfStars);
      event.preventDefault();
    }

    if (event.key === 'ArrowLeft') {
      const ratingMinus = this.rating - this.precision;
      this.rating = clamp(ratingMinus, 0, this.numOfStars);
      event.preventDefault();
    }

    if (event.key === 'Home') {
      this.rating = 0;
      event.preventDefault();
    }

    if (event.key === 'End') {
      this.rating = this.numOfStars;
      event.preventDefault();
    }
  }

  getValueFromMousePosition(event: MouseEvent) {
    const containerLeft = this.element.getBoundingClientRect().left;
    const containerWidth = this.element.getBoundingClientRect().width;

    const numOfSections = this.numOfStars / this.precision;
    const sectionWidth = containerWidth / numOfSections;
    const positionOfMousePointer =
      (event.clientX - containerLeft) / sectionWidth;
    const star = clamp(
      this.roundToPrecision(
        positionOfMousePointer * this.precision,
        this.precision
      ),
      0,
      this.numOfStars
    );
    return star;
  }

  roundToPrecision(numberToRound: number, precision = 1) {
    const multiplier = 1 / precision;
    return Math.ceil(numberToRound * multiplier) / multiplier;
  }

  render() {
    const counter = Array.from(Array(this.numOfStars).keys());
    let displayValue = 0;
    if (!this.interactive || this.disabled) {
      displayValue = this.rating;
    } else {
      displayValue = this.isHovering ? this.hoverValue : this.rating;
    }
    return (
      <div
        class={{
          rating: true,
          'rating--interactive': !this.interactive,
          'rating--disabled': this.disabled,
        }}
        id="rating"
        ref={el => (this.element = el)}
        onMouseMove={this.handleMouseMove}
        onMouseEnter={this.handleMouseEnter}
        onMouseLeave={this.handleMouseLeave}
        onClick={this.handleMouseClick}
        onKeyDown={this.handleKeyDown}
        tabIndex={this.disabled ? -1 : 0}
        aria-label={this.ariaLabel}
      >
        <span class="rating__symbols" onMouseEnter={this.handleMouseEnter}>
          {counter.map(index => (
            <span
              style={{
                width: this.small ? '16px' : '24px',
              }}
              class={{
                rating__symbol: true,
                'rating__symbol--hover':
                  this.isHovering && Math.ceil(displayValue) === index + 1,
              }}
              innerHTML={this.getSymbolBlank()}
              id={`star-${index + 1}`}
            />
          ))}
        </span>
        <span class="rating__symbols rating__symbols--indicator">
          {counter.map(index => (
            <span
              style={{
                width: this.small ? '16px' : '24px',
                clipPath:
                  displayValue > index + 1
                    ? null
                    : `inset(0 ${100 - (displayValue - index) * 100}% 0 0)`,
              }}
              class={{
                rating__symbol: true,
                'rating__symbol--hover':
                  this.isHovering && Math.ceil(displayValue) === index + 1,
              }}
              innerHTML={this.getSymbolFilled()}
            />
          ))}
        </span>
      </div>
    );
  }
}
