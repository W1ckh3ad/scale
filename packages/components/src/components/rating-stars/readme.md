# scale-rating-stars



<!-- Auto Generated Below -->


## Properties

| Property          | Attribute      | Description | Type           | Default                                   |
| ----------------- | -------------- | ----------- | -------------- | ----------------------------------------- |
| `ariaLabel`       | `aria-label`   |             | `string`       | `undefined`                               |
| `disabled`        | `disabled`     |             | `boolean`      | `false`                                   |
| `getSymbolBlank`  | --             |             | `() => string` | `() => '<star-icon></star-icon>'`         |
| `getSymbolFilled` | --             |             | `() => string` | `() => `<star-icon colored></star-icon>`` |
| `hoverValue`      | `hover-value`  |             | `number`       | `0`                                       |
| `interactive`     | `interactive`  |             | `boolean`      | `true`                                    |
| `isHovering`      | `is-hovering`  |             | `boolean`      | `false`                                   |
| `numOfStars`      | `num-of-stars` |             | `number`       | `5`                                       |
| `precision`       | `precision`    |             | `number`       | `0.5`                                     |
| `rating`          | `rating`       |             | `number`       | `0`                                       |
| `small`           | `small`        |             | `boolean`      | `false`                                   |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
