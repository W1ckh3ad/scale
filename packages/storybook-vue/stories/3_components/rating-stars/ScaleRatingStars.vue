<template>
  <scale-rating-stars
    :num-of-stars="numOfStars"
    :rating="rating"
    :small="small"
    :interactive="interactive"
    :disabled="disabled"
    :color-fill="colorFill"
    :aria-translation="ariaTranslation"
  >
  </scale-rating-stars>
</template>

<script>
export default {
  props: {
    numOfStars: { type: Number, default: 5 },
    rating: { type: Number, default: 3 },
    small: { type: Boolean, default: false },
    interactive: { type: Boolean, default: true },
    disabled: { type: Boolean, default: false },
    colorFill: { type: String },
    ariaTranslation: { type: String },
  },
};
</script>
