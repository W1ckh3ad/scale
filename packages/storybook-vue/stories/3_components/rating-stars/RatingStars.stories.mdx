import { Meta, Story, ArgsTable, Canvas } from "@storybook/addon-docs/blocks";
import ScaleRatingStars from "./ScaleRatingStars.vue";

<Meta
  title="Beta Components/Rating Stars"
  component={ScaleRatingStars}
  argTypes={{}}
/>

export const Template = (args, { argTypes }) => ({
  components: { ScaleRatingStars },
  props: {
    label: String,
    ...ScaleRatingStars.props,
  },
  template: `
  <scale-rating-stars
    :num-of-stars="numOfStars"
    :rating="rating"
    :small="small"
    :interactive="interactive"
    :disabled="disabled"
    :color-fill="colorFill"
    :aria-lang="ariaLang"
  >
  </scale-rating-stars>
    `,
});

# Rating Stars

## Standard

<Canvas withSource="none">
  <Story
    name="Standard"
    args={{
      rating: 3,
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

<ArgsTable story="Standard" />

```html
<scale-rating-stars rating="3"></scale-rating-stars>
```

## Disabled

<Canvas withSource="none">
  <Story
    name="Disabled"
    args={{
      disabled: true,
      rating: 2
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-rating-stars rating="3" disabled></scale-rating-stars>
```

## Small

<Canvas withSource="none">
  <Story
    name="Small"
    args={{
      rating: 4,
      small: true
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-rating-stars rating="4" small></scale-rating-stars>
```

## Number of Stars

<Canvas withSource="none">
  <Story
    name="Number of Stars"
    args={{
      rating: 6,
      numOfStars: 10
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-rating-stars rating="6" num-of-stars="10"></scale-rating-stars>
```
