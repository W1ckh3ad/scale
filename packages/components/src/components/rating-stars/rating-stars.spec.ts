import { newSpecPage, SpecPage } from '@stencil/core/testing';
import { RatingStars } from './rating-stars';

async function simulateMouseEvent(
  page: SpecPage,
  strEvent: string,
  elementId: string,
) {
  const event = new MouseEvent(strEvent, {
    view: window,
    bubbles: true,
    cancelable: true,
  });
  const element = page.root.shadowRoot.getElementById(elementId);
  !element.dispatchEvent(event);
}

async function simulateKeyboardEvent(
  page: SpecPage,
  strEvent: string,
  elementId: string,
  key: string
) {
  const event = new KeyboardEvent(strEvent, {
    view: window,
    bubbles: true,
    cancelable: true,
    key: key,
  });
  const element = page.root.shadowRoot.getElementById(elementId);
  !element.dispatchEvent(event);
}

describe('component', () => {
  let page: SpecPage;

  it('builds', () => {
    expect(new RatingStars()).toBeTruthy();
  });
  describe('props', () => {
    beforeEach(async () => {
      page = await newSpecPage({
        components: [RatingStars],
        html: `<scale-rating-stars></scale-rating-stars>`,
      });
    });

    it('has all props', async () => {
      page.root.numOfStars = 7;
      page.root.rating = 3;
      page.root.interactive = true;
      page.root.small = true;
      page.root.color = '#00ff00';
      page.root.ariaLabel = 'rating-stars';
      page.root.precision = 1;
      await page.waitForChanges();
      expect(page.rootInstance.numOfStars).toBe(7);
      expect(page.rootInstance.rating).toBe(3);
      expect(page.rootInstance.interactive).toBe(true);
      expect(page.rootInstance.small).toBe(true);
      expect(page.rootInstance.ariaLabel).toBe('rating-stars');
      expect(page.rootInstance.precision).toBe(1);
      expect(page.rootInstance.hoverValue).toBe(0);
    });

    it('isHovering is set to false on Mouseleave', async () => {
      simulateMouseEvent(page, 'mouseleave', 'rating');
      expect(page.rootInstance.isHovering).toBe(false);
    });

    it('isHovering is set to true on Mouseenter', async () => {
      simulateMouseEvent(page, 'mouseenter', 'rating');
      expect(page.rootInstance.isHovering).toBe(true);
    });

    it('change rating on keydown (ArrowLeft/ArrowRight) ', async () => {
      const precision: number = page.rootInstance.precision;
      expect(await page.rootInstance.rating).toBe(0);
      simulateKeyboardEvent(page, 'keydown', 'rating', 'ArrowRight');
      simulateKeyboardEvent(page, 'keydown', 'rating', 'ArrowRight');
      simulateKeyboardEvent(page, 'keydown', 'rating', 'ArrowLeft');
      const result = 0 + precision + precision - precision;
      expect(await page.rootInstance.rating).toBe(result);
    });
  });

  describe('functions', () => {
    it('roundToPrecision(5, 1) returns 5', () => {
      const component = new RatingStars();
      expect(component.roundToPrecision(5, 1) as any).toBe(5);
    });
    it('handleMouseEnter sets hovering to true', () => {
      const component = new RatingStars();
      component.handleMouseEnter();
      expect(component.isHovering).toBe(true);
    });
    it('handleMouseLeave sets hovering to false', () => {
      const component = new RatingStars();
      component.isHovering = true;
      component.handleMouseLeave();
      expect(component.isHovering).toBe(false);
    });
  });

  describe('snapshots', () => {
    it('rating 0 (initial) should match snapshot', async () => {
      page = await newSpecPage({
        components: [RatingStars],
        html: `<scale-rating-stars></scale-rating-stars>`,
      });
      expect(page.root).toMatchSnapshot();
    });

    it('rating 3 should match snapshot', async () => {
      page = await newSpecPage({
        components: [RatingStars],
        html: `<scale-rating-stars rating="3"></scale-rating-stars>`,
      });
      expect(page.root).toMatchSnapshot();
    });
  });
});

