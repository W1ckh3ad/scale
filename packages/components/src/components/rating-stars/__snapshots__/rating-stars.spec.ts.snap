// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component snapshots rating 0 (initial) should match snapshot 1`] = `
<scale-rating-stars>
  <mock:shadow-root>
    <div class="rating" id="rating" tabindex="0">
      <span class="rating__symbols">
        <span class="rating__symbol" id="star-1" style="width: 24px;">
          <star-icon></star-icon>
        </span>
        <span class="rating__symbol" id="star-2" style="width: 24px;">
          <star-icon></star-icon>
        </span>
        <span class="rating__symbol" id="star-3" style="width: 24px;">
          <star-icon></star-icon>
        </span>
        <span class="rating__symbol" id="star-4" style="width: 24px;">
          <star-icon></star-icon>
        </span>
        <span class="rating__symbol" id="star-5" style="width: 24px;">
          <star-icon></star-icon>
        </span>
      </span>
      <span class="rating__symbols rating__symbols--indicator">
        <span class="rating__symbol" style="width: 24px; clip-path: inset(0 100% 0 0);">
          <star-icon colored=""></star-icon>
        </span>
        <span class="rating__symbol" style="width: 24px; clip-path: inset(0 200% 0 0);">
          <star-icon colored=""></star-icon>
        </span>
        <span class="rating__symbol" style="width: 24px; clip-path: inset(0 300% 0 0);">
          <star-icon colored=""></star-icon>
        </span>
        <span class="rating__symbol" style="width: 24px; clip-path: inset(0 400% 0 0);">
          <star-icon colored=""></star-icon>
        </span>
        <span class="rating__symbol" style="width: 24px; clip-path: inset(0 500% 0 0);">
          <star-icon colored=""></star-icon>
        </span>
      </span>
    </div>
  </mock:shadow-root>
</scale-rating-stars>
`;

exports[`component snapshots rating 3 should match snapshot 1`] = `
<scale-rating-stars rating="3">
  <mock:shadow-root>
    <div class="rating" id="rating" tabindex="0">
      <span class="rating__symbols">
        <span class="rating__symbol" id="star-1" style="width: 24px;">
          <star-icon></star-icon>
        </span>
        <span class="rating__symbol" id="star-2" style="width: 24px;">
          <star-icon></star-icon>
        </span>
        <span class="rating__symbol" id="star-3" style="width: 24px;">
          <star-icon></star-icon>
        </span>
        <span class="rating__symbol" id="star-4" style="width: 24px;">
          <star-icon></star-icon>
        </span>
        <span class="rating__symbol" id="star-5" style="width: 24px;">
          <star-icon></star-icon>
        </span>
      </span>
      <span class="rating__symbols rating__symbols--indicator">
        <span class="rating__symbol" style="width: 24px;">
          <star-icon colored=""></star-icon>
        </span>
        <span class="rating__symbol" style="width: 24px;">
          <star-icon colored=""></star-icon>
        </span>
        <span class="rating__symbol" style="width: 24px; clip-path: inset(0 0% 0 0);">
          <star-icon colored=""></star-icon>
        </span>
        <span class="rating__symbol" style="width: 24px; clip-path: inset(0 100% 0 0);">
          <star-icon colored=""></star-icon>
        </span>
        <span class="rating__symbol" style="width: 24px; clip-path: inset(0 200% 0 0);">
          <star-icon colored=""></star-icon>
        </span>
      </span>
    </div>
  </mock:shadow-root>
</scale-rating-stars>
`;
